### PR TITLE
Feature: Confluent Avro Serializer

### PR DESCRIPTION
**Pull request by lciolecki**
- Add confluent avro serializer

### CODE REVIEW

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroSerializer.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        val out = new ByteArrayOutputStream
+        out.write(AbstractKafkaAvroSerDe.MAGIC_BYTE)
+        out.write(ByteBuffer.allocate(AbstractKafkaAvroSerDe.idSize).putInt(schemaId).array)
+
+        data match {
+          case array: Array[Byte] => out.write(array)
+          case _ =>
+            val encoder = this.encoderFactory.directBinaryEncoder(out, null)
+
+            val value = data match {
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroSerializer.scala:43]: Please extract whole logic of schema evolution to some class like `AvroWriterSchemaEvolutionHandler` or sth similar, with one method `alignRecordToSchema(record, schema)`. We will be able to change implementation and compare if each works the same way.

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroSerializer.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+  /**
+    * Convert serialization record to final schema.
+    * We try to set schema default value when record doesn't support field from schema.
+    *
+    * @param record
+    * @param schema
+    * @return
+    */
+  private def convertRecordToSchema(record: GenericRecord, schema: Schema): GenericData.Record = {
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroSerializer.scala:78]: IMO we should start with implementation that will do serialization -> deserilization (with schema evolution) -> serialization. After that we can do some tweaks if it won't be performed enough. If users will find some bugs, they will be able to switch to correct (slower) implementation. WDYT?

File: build.sbt [hunk 3]
```diff
@@ -145,7 +145,7 @@ val forkSettings = Seq(
 
 val akkaV = "2.5.21" //same version as in Flink
 val flinkV = "1.9.1"
-val avroV = "1.8.2" // should be the same as in flink-avro
+val avroV = "1.9.1" // should be the same as in flink-avro
```
Comment (mproch) [inline build.sbt:148]: what about this comment?

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroSerializer.scala (hunk 2 above)
Comment (mproch) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/serialization/AbstractConfluentKafkaAvroSerializer.scala:78]: :+1: to @arkadius remark, also - don't we need some recursion here if we want it to be correct? 
I would even suggest we leave handling this case (this whole default value business) for the future, and just and remark

File: build.sbt (hunk 3 above)
Comment (lciolecki) [reply build.sbt:148]: `"org.apache.flink" % "flink-avro" % flinkV,` - right now is the same..

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schema/DefaultAvroSchemaEvolution.scala [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        buffer.get(bytes, 0, length)
+        bytes
+      } else {
+        val start = buffer.position + buffer.arrayOffset
+        val binaryDecoder = decoderFactory.binaryDecoder(buffer.array, start, length, null)
+        val result = reader.read(record, binaryDecoder)
+        if (writerSchema.getType == Schema.Type.STRING) result.toString else result
+      }
+    } catch {
+      case exc@(_: RuntimeException | _: IOException) =>
```
Comment (arkadius) [inline engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schema/DefaultAvroSchemaEvolution.scala:69]: Why not to handle Exception instead?

File: build.sbt (hunk 3 above)
Comment (lciolecki) [reply build.sbt:148]: You're right. A will rollback this.

File: engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schema/DefaultAvroSchemaEvolution.scala (hunk 4 above)
Comment (lciolecki) [reply engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schema/DefaultAvroSchemaEvolution.scala:68]: It was copied from confluent AbstractKafkaAvroSerializer ;) They catch only these two exceptions..

**Pull request merged**